### PR TITLE
feature - optimize search

### PR DESCRIPTION
**Pull request by at-tanhle**

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/Library/Ext/ConvertHelperExt.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+//  MyApp
+//
+//  Created by Tanh Le D on 8/30/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import Foundation
+
+class ConverHelper {
+    private static let arrCoDau: [Character] =
```
Comment (at-tuanmai2) [inline MyApp/Library/Ext/ConvertHelperExt.swift]: change name ex: arrSignal

File: MyApp/Library/Ext/ConvertHelperExt.swift [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            "Ê", "Ế", "Ề", "Ể", "Ễ", "Ệ",
+            "Í", "Ì", "Ỉ", "Ĩ", "Ị",
+            "Ó", "Ò", "Ỏ", "Õ", "Ọ",
+            "Ô", "Ố", "Ồ", "Ổ", "Ỗ", "Ộ",
+            "Ơ", "Ớ", "Ờ", "Ở", "Ỡ", "Ợ",
+            "Ú", "Ù", "Ủ", "Ũ", "Ụ",
+            "Ư", "Ứ", "Ừ", "Ử", "Ữ", "Ự",
+            "Ý", "Ỳ", "Ỷ", "Ỹ", "Ỵ"]
+
+    private static let arrKhongDau: [Character] =
```
Comment (at-tuanmai2) [inline MyApp/Library/Ext/ConvertHelperExt.swift]: change name ex: arrNoSignal

File: MyApp/Model/Schema/MenuItem.swift [hunk 3]
```diff
@@ -13,11 +13,17 @@ import RealmSwift
...
+    dynamic var name = "" {
+        didSet {
+            convertedName = ConverHelper.convertVietNam(text: name)
+        }
+    }
     dynamic var imagePath = ""
     dynamic var price = 0
     dynamic var itemDescription = ""
 
+    dynamic var convertedName = ""
```
Comment (at-tuanmai2) [inline MyApp/Model/Schema/MenuItem.swift]: change name: keyWord chẳng hạn

File: MyApp/Model/Schema/MenuItem.swift [hunk 4]
```diff
@@ -36,7 +42,12 @@ final class MenuItem: Object, StaticMappable {
         return RealmS().object(ofType: self, forMapping: map)
     }
 
-    static func query() -> Results<MenuItem>? {
+    static func query() -> Results<MenuItem> {
         return RealmS().objects(self)
     }
+
+    static func query(key: String) -> Results<MenuItem> {
+        let predicate = NSPredicate(format: "convertedName CONTAINS [c] %@", key)
```
Comment (at-tuanmai2) [inline MyApp/Model/Schema/MenuItem.swift]: please define query format

File: MyApp/View/Base/ViewController.swift [hunk 5]
```diff
@@ -28,36 +28,37 @@ class ViewController: UIViewController, MVVM.View {
...
         // back button
         let backButton = UIButton(frame: CGRect(x: 0, y: 0, width: 20 * ratio, height: 20 * ratio))
         backButton.setBackgroundImage(#imageLiteral(resourceName: "back"), for: .normal)
         backButton.addTarget(self, action: #selector(backButtonTouchUpInside), for: .touchUpInside)
-        backButton.addSubview(backLabel)
 
+        // back label
+        let backLabel = UILabel()
+        backLabel.frame = CGRect(x: 25, y: backButton.frame.origin.y, width: 150 * ratio, height: 20 * ratio)
+        backLabel.text = "Quay lại"
```
Comment (at-tuanmai2) [inline MyApp/View/Base/ViewController.swift]: defile text

File: MyApp/View/Cell/SearchHistoryCell.swift [hunk 6]
```diff
@@ -29,6 +30,7 @@ class SearchHistoryCell: UITableViewCell, MVVM.View {
 
     func updateView() {
         guard let viewModel = viewModel else { return }
-        textLabel?.text = viewModel.keyWord
+        let dateString = DateFormatter.dateTime.string(from: viewModel.dateSearch as Date)
```
Comment (at-tuanmai2) [inline MyApp/View/Cell/SearchHistoryCell.swift]: viewmodel

File: MyApp/View/Cell/SearchHistoryCell.swift [hunk 7]
```diff
@@ -29,6 +30,7 @@ class SearchHistoryCell: UITableViewCell, MVVM.View {
...
+        textLabel?.text = viewModel.keyWord + "   " + dateString
```
Comment (at-tuanmai2) [inline MyApp/View/Cell/SearchHistoryCell.swift]: viewmodel.....

File: MyApp/View/Controllers/Search/SearchViewController.swift [hunk 8]
```diff
@@ -21,31 +23,55 @@ final class SearchViewController: ViewController {
...
+            self.view.addSubview(searchHistoryView)
+            self.tabBarController?.tabBar.isHidden = true
+        })
+    }
+
+    func searchHistoryViewDisappear() {
+        guard let searchHistoryView = searchHistoryView else { return }
+        UIView.transition(with: view, duration: 0.3, options: .transitionCrossDissolve, animations: {
+            searchHistoryView.removeFromSuperview()
+            self.searchBar.showsCancelButton = false
```
Comment (at-tuanmai2) [inline MyApp/View/Controllers/Search/SearchViewController.swift]: showsCancelButton?

File: MyApp/View/Controllers/Search/SearchViewController.swift [hunk 9]
```diff
@@ -88,25 +114,67 @@ extension SearchViewController: UICollectionViewDelegateFlowLayout {
...
 extension SearchViewController: UISearchBarDelegate {
     func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
-        guard let searchHistoryView = searchHistoryView else { return }
-        UIView.transition(with: view, duration: 0.3, options: .transitionCrossDissolve, animations: {
-            searchBar.showsCancelButton = true
-            self.view.addSubview(searchHistoryView)
-            self.tabBarController?.tabBar.isHidden = true
-        })
+        isSearchHistoryVisible = true
+        searchHistoryViewAppear()
```
Comment (at-tuanmai2) [inline MyApp/View/Controllers/Search/SearchViewController.swift]: searchHistoryViewAppear -> searchHistoryViewWillAppear or searchHistoryViewDidAppear

File: MyApp/View/Controllers/Search/SearchViewController.swift [hunk 10]
```diff
@@ -88,25 +114,67 @@ extension SearchViewController: UICollectionViewDelegateFlowLayout {
...
     }
 
     func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+        isSearchHistoryVisible = false
+        searchHistoryViewDisappear()
```
Comment (at-tuanmai2) [inline MyApp/View/Controllers/Search/SearchViewController.swift]: same

**Pull request merged**